brian2/codegen/runtime/weave_rt/templates/spatialstateupdate.cpp

```diff
@@ -5,22 +5,28 @@
 {# USES_VARIABLES { Cm, dt, v, N,
                   ab_star0, ab_star1, ab_star2, b_plus,
                   ab_plus0, ab_plus1, ab_plus2, b_minus,
-                  ab_minus0, ab_minus1, ab_minus2, v_star, u_plus, u_minus} #}
+                  ab_minus0, ab_minus1, ab_minus2, v_star, u_plus, u_minus,
+                  _P, _B, _morph_i, _morph_parent_i, _starts, _ends,
+                  _invr0, _invrn} #}
 
 {% macro main() %}
     {{ common.insert_group_preamble() }}
 
 	double *_gtot_all=(double *)malloc(N*sizeof(double));
-	double *c=(double *)malloc(N*sizeof(double));
+	double *c=(double *)calloc(N, sizeof(double));
 	double ai,bi,_m;

```

Comment:
I'd rather avoid having such internal variables as state variables, they are an implementation detail and only used within the template (they are not used in numpy, for example). But we could do the allocation only once by declaring the pointers as static and doing the allocation only if they haven't been allocated before. The only problem I'm seeing with that is that we can't de-allocate the memory. For standard runs this is certainly not an issue, but if a user creates multi-compartmental neurons in a loop, this would be a memory leak. In the interest of keeping things simple, I'd therefore tend to keep them as they are. I don't quite remember why I changed the `malloc` into the `calloc`, though, looking at the code it seems to be never used before it gets a value assigned.